dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DefaultDeduplicationService.java

```diff
@@ -101,6 +105,70 @@ public void updatePotentialDuplicate( PotentialDuplicate potentialDuplicate )
         potentialDuplicateStore.update( potentialDuplicate );
     }
 
+    @Override
+    public boolean isAutoMergeable( PotentialDuplicate potentialDuplicate )
+    {
+        TrackedEntityInstance trackedEntityInstanceA = Optional.ofNullable( trackedEntityInstanceService
+            .getTrackedEntityInstance( potentialDuplicate.getTeiA() ) )
+            .orElseThrow( () -> new PotentialDuplicateException(
+                "No tracked entity instance found with id '" + potentialDuplicate.getTeiA() + "'." ) );
+
+        TrackedEntityInstance trackedEntityInstanceB = Optional.ofNullable( trackedEntityInstanceService
+            .getTrackedEntityInstance( potentialDuplicate.getTeiB() ) )
+            .orElseThrow( () -> new PotentialDuplicateException(
+                "No tracked entity instance found with id '" + potentialDuplicate.getTeiB() + "'." ) );
+
+        if ( !trackedEntityInstanceA.getTrackedEntityType().equals( trackedEntityInstanceB.getTrackedEntityType() ) )
+        {
+            return false;
+        }
+
+        if ( trackedEntityInstanceA.isDeleted() || trackedEntityInstanceB.isDeleted() )
+        {
+            return false;
+        }
+
+        if ( trackedEntityInstanceA.getRelationshipItems().stream()
+            .anyMatch( relationshipItem -> null != relationshipItem.getRelationship().getTo().getTrackedEntityInstance()
+                && relationshipItem.getRelationship().getTo().getTrackedEntityInstance()
+                    .equals( trackedEntityInstanceB ) ) )
+        {
+            return false;
+        }
+
+        if ( trackedEntityInstanceB.getRelationshipItems().stream()
+            .anyMatch( relationshipItem -> null != relationshipItem.getRelationship().getTo().getTrackedEntityInstance()
+                && relationshipItem.getRelationship().getTo().getTrackedEntityInstance()
+                    .equals( trackedEntityInstanceA ) ) )
+        {
+            return false;
+        }
+
+        Set<TrackedEntityAttributeValue> trackedEntityAttributeValueA = trackedEntityInstanceA
+            .getTrackedEntityAttributeValues();
+        Set<TrackedEntityAttributeValue> trackedEntityAttributeValueB = trackedEntityInstanceB
+            .getTrackedEntityAttributeValues();
+
+        if ( null == trackedEntityAttributeValueA || null == trackedEntityAttributeValueB )
```

Comment:
As far as I can see, these variables are initialized when the object is constructed, so maybe isEmpty() is better?

Reply:
i'll add it and leave both